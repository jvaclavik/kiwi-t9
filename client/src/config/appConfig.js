// @flow
export default {
  apiUrl: "http://localhost:3000",
  onNumberPressDelay: 300,
}
