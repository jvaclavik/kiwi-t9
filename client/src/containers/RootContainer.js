// @flow
import React from "react"
import { SafeAreaView } from "react-native"
import debounce from "lodash/debounce"

// config
import appConfig from "../config/appConfig"

// components
import { Message, Keyboard, WordList, Title } from "../components"

// utils
import renderStyle from "../utils/renderStyle"

// styles
import { Colors } from "../themes"

// types
import { Api } from "../types"

const styles = {
  container: {
    flex: 1,
    backgroundColor: Colors.background,
  },
  title: {
    height: 40,
  },
  wordList: {
    height: 44,
  },
  message: {
    flex: 1,
  },
  keyboard: {
    flex: 3,
    alignItems: "center",
    justifyContent: "center",
  },
}

type Props = {|
  +api: Api,
|}

const defaultState = {
  input: "",
  message: "",
  wordList: [],
  loading: false,
}

export default class RootContainer extends React.PureComponent<Props> {
  state = defaultState

  debouncedInput = debounce(() => {
    const { input } = this.state
    this.getWordsFromApi(input)
  }, appConfig.onNumberPressDelay)

  getWordsFromApi = input => {
    const { api } = this.props
    this.setState({ loading: true })
    api
      .getWords(input)
      .then(response => response.json())
      .then(data => {
        this.setState({ loading: false, wordList: data })
      })
      .catch(error => {
        console.log(error) // @TODO handle error
      })
  }

  onBackspacePress = () => {
    const { input } = this.state
    if (input.length > 1) {
      this.setState(prevState => ({
        input: prevState.input.slice(0, -1),
      }))
      this.debouncedInput()
    } else {
      this.setState({
        input: "",
        wordList: [],
      })
    }
  }

  onResetPress = () => {
    this.setState(defaultState)
  }

  onNumberPress = value => {
    switch (value) {
      case "0":
        this.setState(prevState => ({
          input: "",
          wordList: [],
          message: `${prevState.message}${
            prevState.wordList.length > 0 ? prevState.wordList[0] : ""
          } `,
        }))
        break
      case "1":
        break
      default:
        this.setState(prevState => ({ input: `${prevState.input}${value}` }))
        this.debouncedInput()
    }
  }

  onWordPress = word => {
    this.setState(prevState => ({
      input: "",
      wordList: [],
      message: `${prevState.message}${word} `,
    }))
  }

  render() {
    const { message, wordList, loading, input } = this.state
    return (
      <SafeAreaView style={renderStyle(styles.container)}>
        <Title style={renderStyle(styles.title)}>Kiwi T9</Title>
        <Message style={renderStyle(styles.message)} message={message} />
        <WordList
          loading={loading}
          onWordPress={this.onWordPress}
          words={wordList}
          style={renderStyle(styles.wordList)}
        />
        <Keyboard
          showReset={message || input}
          showBackspace={input}
          onNumberPress={this.onNumberPress}
          onBackspacePress={this.onBackspacePress}
          onResetPress={this.onResetPress}
          style={renderStyle(styles.keyboard)}
        />
      </SafeAreaView>
    )
  }
}
