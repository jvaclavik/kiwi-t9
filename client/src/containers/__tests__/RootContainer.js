// @flow
import React from "react"
import renderer from "react-test-renderer"
import fetch from "isomorphic-fetch"

// services
import createApi from "../../services/api"

// config
import appConfig from "../../config/appConfig"

// components
import RootContainer from "../RootContainer"

test("renders correctly with required props", () => {
  const tree = renderer
    .create(<RootContainer api={createApi(fetch, appConfig)} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
