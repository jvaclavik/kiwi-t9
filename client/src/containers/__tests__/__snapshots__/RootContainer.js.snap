// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with required props 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "height": 40,
        "marginHorizontal": 40,
        "marginVertical": 10,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#000",
          "fontSize": 32,
          "fontWeight": "900",
        }
      }
    >
      Kiwi T9
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <View
    style={
      Object {
        "height": 44,
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 3,
        "justifyContent": "center",
      }
    }
  />
</RCTSafeAreaView>
`;
